PID controller and diff drive controller chaining (#710)
ros-controls/ros2_control_demos: example_16/doc/userdoc.rst

:github_url: https://github.com/ros-controls/ros2_control_demos/blob/{REPOS_FILE_BRANCH}/example_16/doc/userdoc.rst

.. _ros2_control_demos_example_16_userdoc:

********************************
DiffBot with Chained Controllers
********************************

*DiffBot*, or ''Differential Mobile Robot'', is a simple mobile base with differential drive. The robot is basically a box moving according to differential drive kinematics.

This example extends *example_2* by demonstrating controller chaining with a diff_drive_controller and two pid_controllers (one for each wheel) to achieve coordinated robot motion. If you haven't already, you can find the instructions for *example_2* in :ref:`ros2_control_demos_example_2_userdoc`. It is recommended to follow the steps given in that tutorial first before proceeding with this one.

> This example demonstrates controller chaining as described in the `ROS2 controller manager chaining documentation <https://github.com/ros-controls/ros2_control/blob/master/controller_manager/doc/controller_chaining.rst?plain=1>`__. The control chain flows from the diff_drive_controller through two PID controllers to the DiffBot hardware. The diff_drive_controller converts desired robot twist into wheel velocity commands, which are then processed by the PID controllers to directly control the wheel velocities. Additionally, this example shows how to enable the feedforward mode for the PID controllers.

Reviewer: ```suggestion
This example demonstrates controller chaining as described in :ref:`controller_chaining`. The control chain flows from the diff_drive_controller through two PID controllers to the DiffBot hardware. The diff_drive_controller converts desired robot twist into wheel velocity commands, which are then processed by the PID controllers to directly control the wheel velocities. Additionally, this example shows how to enable the feedforward mode for the PID controllers.
```

      time_step: 0.099941354
      output: 32.090205119481226

Visualize the convergence of DiffBot's wheel velocities and commands
---------------------------------------------------------------------

In the section below, we will use *plotjuggler* to observe the convergence of DiffBot's wheel velocities and commands from PID controllers.

*plotjuggler* is an open-source data visualization tool and widely embraced by ROS2 community. If you don't have it installed, you can find the instructions from `plotjuggler website <https://github.com/facontidavide/PlotJuggler>`__.

Before we proceed, we stop all previous steps from terminal and start from the beginning.

1. To start *DiffBot* example open a terminal, source your ROS2-workspace and execute its launch file with

   .. code-block:: shell

    ros2 launch ros2_control_demo_example_16 diffbot.launch.py

Like before, if you can see an orange box in *RViz*, everything has started properly.

2. To start the plotjuggler with a provided layout file(plotjuggler.xml), open another terminal and run following command.

   .. code-block:: shell

    # replace {PARENT_DIR_OF_LAYOUT_FILE} with actual parent directory of the layout file

>     ros2 run plotjuggler plotjuggler --layout {PARENT_DIR_OF_LAYOUT_FILE}/plotjuggler.xml

Reviewer: ```suggestion
    ros2 run plotjuggler plotjuggler --layout $(ros2 pkg prefix ros2_control_demo_example_16 --share)/config/plotjuggler.xml
```

>     # replace {PARENT_DIR_OF_LAYOUT_FILE} with actual parent directory of the layout file
>     ros2 run plotjuggler plotjuggler --layout {PARENT_DIR_OF_LAYOUT_FILE}/plotjuggler.xml

Reviewer: ```suggestion
    ros2 run plotjuggler plotjuggler --layout $(ros2 pkg prefix ros2_control_demo_example_16 --share)/config/plotjuggler.xml
```

After this, you will see a few dialogs popping up. For example:

   .. code-block:: shell

    Start the previously used streaming plugin?

    ROS2 Topic Subscriber

> Click 'Yes' for the first dialog and 'OK" to the follow two dialogs, then you will see the plotjuggler window.

Reviewer: ```suggestion
Click 'Yes' for the first dialog and 'OK" to the following two dialogs, then you will see the plotjuggler window.
```

3. To enable feedforward mode and published a command to move the robot, instead of doing these manually, we will use the demo_test.launch.py. Open another terminal and execute

   .. code-block:: shell

    ros2 launch ros2_control_demo_example_16 demo_test.launch.py

4. From the plotjuggler, you can see the controllers' states and commands being plotted, similar to following figure. From the figure, the DiffBot's wheel velocities and commands from PID controllers are converged to the target velocity fairly quickly.

>   .. image:: diffbot_velocities.png
>     :width: 400
>     :alt: Plotjuggler visualization of Diffbot velocities and commands
> 
>   TODO: add png here.
> 

Reviewer: ```suggestion
  .. image:: diffbot_velocities.png
    :width: 400
    :alt: Plotjuggler visualization of DiffBot velocities and commands

```

> 5. Change the ``gains`` in the ``diffbot_chained_controllers.yaml`` file with different p, i, d values, repeat above steps to see the effect of feedforward mode.

Reviewer: PID controller supports online reconfiguration of the gains, e.g. with ros2cli or rqt_reconfigure. Not sure if we want to add this as a comment here

Author: Done, thank you @christophfroehlich  for many helpful suggestions. These are good learning for me.
